server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2065,6 +2065,12 @@ public void startRecovery(RecoveryState recoveryState, PeerRecoveryTargetService
         }
     }
 
+    public ShardStateMetaData getShardStateMetaData() {
+        synchronized (mutex) {
```

Comment:
we can avoid the mutex here. just do a one-time volatile read of shardrouting (which is an immutable object). `indexSettings.getUUID()` are a final object and the uuid is immutable.